### PR TITLE
CB-12410 fix repair for HA clusters (Medium Duty DL + HA DH)

### PR DESCRIPTION
**Pull request by sodre90**
- All gateway instances registered to cluster proxy with it’s instanceid and re-registered during repair.
- Salt communication happens to all salt masters using the unique cluster proxy service endpoint.
- Attach volume to the appropriate instance with the same FQDN at attaching phase
- Doc for this solution in cloudera drive: "Datalake/Datahub SaltStack HA operations"
- Manually tested cases: 
   - Start Medium Duty DL and HA DH clusters without fix on Azure and AWS side, then start cloudbreak with this change and do repair for at least 2 hostgroups on them. 
   - Tested if GW hostgroup is healthy but idbroker is unhealthy, then repair.
   - Start DL without fix, start cloudbreak with this change, then do upgrade
- Automated tests:
   - Medium duty DL repair idbroker and GW hostgroup
   - Medium duty DL upgrade

### CODE REVIEW

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/service/ClusterProxyService.java [hunk 1]
```diff
@@ -135,10 +137,26 @@ private ConfigRegistrationRequest createProxyConfigReRegisterRequest(Stack stack
...
+        String primaryGWInternalAdminUrl = internalAdminUrl(stack, ServiceFamilies.GATEWAY.getDefaultPort());
+        LOGGER.info("Primary GW internal admin URL is: {}", primaryGWInternalAdminUrl);
+        List<ClusterServiceConfig> clusterServiceConfigs = getClusterServiceConfigsForGWs(stack);
+        clusterServiceConfigs.add(cmServiceConfig(stack, null, "cloudera-manager", clusterManagerUrl(stack)));
+        clusterServiceConfigs.add(cmServiceConfig(stack, clientCertificates(stack), CB_INTERNAL, primaryGWInternalAdminUrl));
+        LOGGER.info("Service configs: {}", clusterServiceConfigs);
+        return clusterServiceConfigs;
+    }
+
+    private List<ClusterServiceConfig> getClusterServiceConfigsForGWs(Stack stack) {
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/service/ClusterProxyService.java:149]: bit of refactor:
```
    private List<ClusterServiceConfig> getClusterServiceConfigsForGWs(Stack stack) {
        List<InstanceMetaData> gatewayInstanceMetadatas = stack.getGatewayInstanceMetadata();
        return gatewayInstanceMetadatas.stream()
                .map(gatewayInstanceMetadata -> createClusterServiceConfigFromGWMetadata(stack, gatewayInstanceMetadata))
                .collect(Collectors.toList());
    }

    private ClusterServiceConfig createClusterServiceConfigFromGWMetadata(Stack stack, InstanceMetaData gatewayInstanceMetadata) {
        String gatewayIp = gatewayInstanceMetadata.getPublicIpWrapper();
        String internalAdminUrl = String.format("https://%s:%d", gatewayIp, ServiceFamilies.GATEWAY.getDefaultPort());
        return cmServiceConfig(stack, clientCertificates(stack),
                CB_INTERNAL + "-" + gatewayInstanceMetadata.getInstanceId(), internalAdminUrl);
    }

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/service/ClusterProxyService.java [hunk 2]
```diff
@@ -135,10 +137,26 @@ private ConfigRegistrationRequest createProxyConfigReRegisterRequest(Stack stack
...
+        List<ClusterServiceConfig> clusterServiceConfigs = new ArrayList<>();
+        List<InstanceMetaData> gatewayInstanceMetadatas = stack.getGatewayInstanceMetadata();
+        for (InstanceMetaData gatewayInstanceMetadata : gatewayInstanceMetadatas) {
+            String gatewayIp = gatewayInstanceMetadata.getPublicIpWrapper();
+            String internalAdminUrl = String.format("https://%s:%d", gatewayIp, ServiceFamilies.GATEWAY.getDefaultPort());
+            ClusterServiceConfig gatewayClusterServiceConfig = cmServiceConfig(stack, clientCertificates(stack),
+                    CB_INTERNAL + "-" + gatewayInstanceMetadata.getInstanceId(), internalAdminUrl);
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/service/ClusterProxyService.java:156]: I guess this would be different from freeipa solution, where we use the fqdn. Is this because we don't have it at this time?

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/service/ClusterProxyService.java [hunk 3]
```diff
@@ -225,8 +243,16 @@ private String internalAdminUrl(Stack stack, int port) {
         return String.format("https://%s:%d", gatewayIp, port);
     }
 
-    public String getProxyPath(String crn) {
-        return String.format("%s/proxy/%s/%s", clusterProxyConfiguration.getClusterProxyBasePath(), crn, CB_INTERNAL);
+    public String getProxyPath(String crn, String internalIdentity) {
+        ReadConfigResponse readConfigResponse = clusterProxyRegistrationClient.readConfig(crn);
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/service/ClusterProxyService.java:247]: while I understand the motivation of checking this, but this could lead to extra load on CP side as well as adding delay to client building. I can see 2 way for improving this:
- store locally when we register this, then we don't have to call CP
- cache the response to avoid calls

what do you think?

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/service/ClusterProxyService.java [hunk 4]
```diff
@@ -225,8 +243,16 @@ private String internalAdminUrl(Stack stack, int port) {
         return String.format("https://%s:%d", gatewayIp, port);
     }
 
-    public String getProxyPath(String crn) {
-        return String.format("%s/proxy/%s/%s", clusterProxyConfiguration.getClusterProxyBasePath(), crn, CB_INTERNAL);
+    public String getProxyPath(String crn, String internalIdentity) {
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/service/ClusterProxyService.java:246]: could you please add logging? also I would move out the decesion logic to a separate method:
```
    public String getProxyPath(String crn, String internalIdentity) {
        String endpointWithIdentity = CB_INTERNAL + "-" + internalIdentity;
        if (isServiceEndpointWithIdentityRegistered(crn, endpointWithIdentity)) {
            <LOG>
            return String.format("%s/proxy/%s/%s", clusterProxyConfiguration.getClusterProxyBasePath(), crn, endpointWithIdentity);
        } else {
            <LOG>
            return String.format("%s/proxy/%s/%s", clusterProxyConfiguration.getClusterProxyBasePath(), crn, CB_INTERNAL);
        }
    }

    private boolean isServiceEndpointWithIdentityRegistered(String crn, String endpointWithIdentity) {
        ReadConfigResponse readConfigResponse = clusterProxyRegistrationClient.readConfig(crn);
        return readConfigResponse.getServices().stream()
                .anyMatch(readConfigService -> endpointWithIdentity.equals(readConfigService.getName()));
    }

File: core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ClusterRepairService.java [hunk 5]
```diff
@@ -192,6 +194,21 @@ public FlowIdentifier repairNodes(Long stackId, Set<String> nodeIds, boolean del
...
+        }
+        if (ManualClusterRepairMode.HOST_GROUP.equals(repairMode)) {
+            return stack.getPrimaryGatewayInstance().isUnhealthy() &&
+                    !selectedParts.contains(stack.getPrimaryGatewayInstance().getInstanceGroup().getGroupName());
+        } else if (ManualClusterRepairMode.NODE_ID.equals(repairMode)) {
+            return stack.getPrimaryGatewayInstance().isUnhealthy() &&
+                    !selectedParts.contains(stack.getPrimaryGatewayInstance().getInstanceId());
+        } else {
+            return false;
+        }
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ClusterRepairService.java:209]: coudl you please cover this with unit tests?

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/InstanceMetaDataService.java [hunk 6]
```diff
@@ -55,14 +56,19 @@ public void updateInstanceStatus(final InstanceMetaData instanceMetaData, com.se
...
             InstanceGroup instanceGroup = getInstanceGroup(stack.getInstanceGroups(), cloudInstance.getTemplate().getGroupName());
             if (instanceGroup != null) {
                 InstanceMetaData instanceMetaData = new InstanceMetaData();
                 instanceMetaData.setPrivateId(cloudInstance.getTemplate().getPrivateId());
                 instanceMetaData.setInstanceStatus(com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.base.InstanceStatus.REQUESTED);
                 instanceMetaData.setInstanceGroup(instanceGroup);
+                if (hostNameIterator.hasNext()) {
+                    String hostName = hostNameIterator.next();
+                    instanceMetaData.setDiscoveryFQDN(hostName);
+                }
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/InstanceMetaDataService.java:71]: it would be nice to have logs here to see when we set hostnames, at least for now

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java [hunk 7]
```diff
@@ -153,14 +158,24 @@ public int saveInstanceMetaData(Stack stack, Iterable<CloudVmMetaDataStatus> clo
...
-                                instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.GATEWAY);
+                                if (!primaryIgSelected) {
+                                    primaryIgSelected = true;
+                                    instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.GATEWAY_PRIMARY);
+                                    instanceMetaDataEntry.setServer(Boolean.TRUE);
+                                    LOGGER.info("Primary gateway is not selected, let's select this instance: {}", instanceMetaDataEntry.getInstanceId());
+                                } else {
+                                    LOGGER.info("Primary gateway was selected");
+                                    instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.GATEWAY);
+                                }
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java:178]: could you please do a bit refactor and move at least these out to methods? this method is already large as hell

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java [hunk 8]
```diff
@@ -153,14 +158,24 @@ public int saveInstanceMetaData(Stack stack, Iterable<CloudVmMetaDataStatus> clo
                 if (instanceMetaDataEntry.getInstanceMetadataType() == null) {
                     if (ig != null) {
                         if (InstanceGroupType.GATEWAY.equals(ig.getInstanceGroupType())) {
-                            if (!primaryIgSelected) {
-                                primaryIgSelected = true;
-                                instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.GATEWAY_PRIMARY);
-                                instanceMetaDataEntry.setServer(Boolean.TRUE);
-                                LOGGER.info("Primary gateway is not selected, let's select this instance: {}", instanceMetaDataEntry.getInstanceId());
+                            if (terminatedPrimaryGwWhichShouldBeRestored.isPresent()) {
+                                if (terminatedPrimaryGwWhichShouldBeRestored.get().getDiscoveryFQDN().equals(instanceMetaDataEntry.getDiscoveryFQDN())) {
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java:162]: is it a possible scenario we would have an already terminated PGW but don't have IM with the same FQDN or we would have IM with empty FQDN?
If yes this code would leave us without PGW

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java (hunk 7 above)
Comment (lacikaaa) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java:178]: please cover these with unit tests

File: orchestrator-salt/src/main/java/com/sequenceiq/cloudbreak/orchestrator/salt/SaltOrchestrator.java [hunk 9]
```diff
@@ -949,7 +949,7 @@ public void stopClusterManagerAgent(GatewayConfig gatewayConfig, Set<Node> allNo
                 Set<String> allHostnames = responsiveNodesUnderStopping.stream().map(Node::getHostname).collect(Collectors.toSet());
                 runSyncAll(sc, allHostnames, responsiveNodesUnderStopping, exitCriteriaModel);
 
-                refreshPillars(gatewayConfig, responsiveNodes, exitCriteriaModel, sc);
+                refreshPillars(gatewayConfig, allNodes, exitCriteriaModel, sc);
```
Comment (lacikaaa) [inline orchestrator-salt/src/main/java/com/sequenceiq/cloudbreak/orchestrator/salt/SaltOrchestrator.java:952]: why was this change necessary? could you add the explanation to the commit message?

**Review by lacikaaa — CHANGES_REQUESTED** [review]
please write an extensive commit message about:
- what and why was changed
- what are the cases you have manually tested
- what are the cases covered by e2e or int tests

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/service/ClusterProxyService.java (hunk 2 above)
Comment (sodre90) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/service/ClusterProxyService.java:156]: Yes, unfortunately we don't have FQDN at this time :( My first implementation used FQDN but it failed. If somebody moves the FQDN generation part to the beginning of the stack creation then we can rewrite it, but I didn't want to rewrite that part also.

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/service/ClusterProxyService.java (hunk 3 above)
Comment (sodre90) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/service/ClusterProxyService.java:247]: I think CP does only a few things, so I really hope it can read it very fast. But if I could choose between storing it or caching I would go with storing this. Please discuss in private.

File: cloud-template/src/main/java/com/sequenceiq/cloudbreak/cloud/template/context/VolumeMatcher.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+    public static void addVolumeResourcesToContext(List<CloudInstance> instances, List<CloudResource> groupInstances, List<CloudResource> groupVolumeSets,
+            ResourceBuilderContext context) {
+        List<CloudResource> groupVolumeSetsWithFQDN = groupVolumeSets.stream()
+                .filter(groupVolumeSet -> groupVolumeSet.getParameter("attributes", VolumeSetAttributes.class).getDiscoveryFQDN() != null)
+                .collect(Collectors.toList());
+        LOGGER.info("Volume sets with FQDN: " + groupVolumeSetsWithFQDN);
+        List<CloudResource> groupVolumeSetsWithoutFQDN = groupVolumeSets.stream()
+                .filter(groupVolumeSet -> groupVolumeSet.getParameter("attributes", VolumeSetAttributes.class).getDiscoveryFQDN() == null)
+                .collect(Collectors.toList());
```
Comment (lnardai) [inline cloud-template/src/main/java/com/sequenceiq/cloudbreak/cloud/template/context/VolumeMatcher.java:29]: You could do the initial splitting of VolumeSets with/withOut FQDN with java streams partition. In high node count cases it might even perform better.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java [hunk 11]
```diff
@@ -153,14 +158,24 @@ public int saveInstanceMetaData(Stack stack, Iterable<CloudVmMetaDataStatus> clo
...
                             } else {
-                                LOGGER.info("Primary gateway was selected");
-                                instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.GATEWAY);
+                                if (!primaryIgSelected) {
+                                    primaryIgSelected = true;
+                                    instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.GATEWAY_PRIMARY);
+                                    instanceMetaDataEntry.setServer(Boolean.TRUE);
+                                    LOGGER.info("Primary gateway is not selected, let's select this instance: {}", instanceMetaDataEntry.getInstanceId());
+                                } else {
+                                    LOGGER.info("Primary gateway was selected");
```
Comment (lnardai) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java:176]: Can we correct this logging to:
"Primary gateway already selected. Using  instanceMetadata type Gateway for instance: {}"

Or somehow harmonize with the orher else branch:

`                                    LOGGER.info("Primary gateway is another instance not this one: {}", instanceMetaDataEntry.getInstanceId());`

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java [hunk 12]
```diff
@@ -153,14 +158,24 @@ public int saveInstanceMetaData(Stack stack, Iterable<CloudVmMetaDataStatus> clo
                 if (instanceMetaDataEntry.getInstanceMetadataType() == null) {
                     if (ig != null) {
                         if (InstanceGroupType.GATEWAY.equals(ig.getInstanceGroupType())) {
-                            if (!primaryIgSelected) {
-                                primaryIgSelected = true;
-                                instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.GATEWAY_PRIMARY);
-                                instanceMetaDataEntry.setServer(Boolean.TRUE);
-                                LOGGER.info("Primary gateway is not selected, let's select this instance: {}", instanceMetaDataEntry.getInstanceId());
+                            if (terminatedPrimaryGwWhichShouldBeRestored.isPresent()) {
```
Comment (lnardai) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java:161]: If I want a pseudocode from what happens here is (I think):
1. we first would collect all the GATEWAY or GATEWAY_PRIMARY nodes
2. in case it's a repair and we are repairing a PGW we don't change anything 
3. if we found no PWG we elect the first one

If so can we reorganise the code according to this logic? The now 5 ifs makes this quite hard to understand and log properly.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/datalake/upgrade/handler/DatalakeVmReplaceWaitHandler.java [hunk 13]
```diff
@@ -32,7 +32,7 @@
     @Value("${sdx.vm.replace.sleeptime_sec:20}")
     private int sleepTimeInSec;
 
-    @Value("${sdx.vm.replace.duration_min:60}")
+    @Value("${sdx.vm.replace.duration_min:120}")
```
Comment (lnardai) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/datalake/upgrade/handler/DatalakeVmReplaceWaitHandler.java:35]: Is this 120 based on something, or just the usual make it double? :)

**Review by lnardai — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java (hunk 8 above)
Comment (sodre90) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java:162]: Good catch, I will cover this case.

File: orchestrator-salt/src/main/java/com/sequenceiq/cloudbreak/orchestrator/salt/SaltOrchestrator.java (hunk 9 above)
Comment (sodre90) [reply orchestrator-salt/src/main/java/com/sequenceiq/cloudbreak/orchestrator/salt/SaltOrchestrator.java:952]: It is resolved by @lturcsanyi, so I will rebase to the latest master and it will disappear

File: cloud-template/src/main/java/com/sequenceiq/cloudbreak/cloud/template/context/VolumeMatcher.java (hunk 10 above)
Comment (sodre90) [reply cloud-template/src/main/java/com/sequenceiq/cloudbreak/cloud/template/context/VolumeMatcher.java:29]: Yes, it can be faster but it will cause a higher load on a single pod and in a multithreaded environment I am not a huge fan of stream partition because it will consume performance from other parallel tasks/ queries.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/datalake/upgrade/handler/DatalakeVmReplaceWaitHandler.java (hunk 13 above)
Comment (sodre90) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/datalake/upgrade/handler/DatalakeVmReplaceWaitHandler.java:35]: it is based on the case medium duty template contains 5 host groups and we do upgrade group by group :D

File: cloud-template/src/main/java/com/sequenceiq/cloudbreak/cloud/template/context/VolumeMatcher.java (hunk 10 above)
Comment (sodre90) [reply cloud-template/src/main/java/com/sequenceiq/cloudbreak/cloud/template/context/VolumeMatcher.java]: I asked CP guys and it is okay to call CP readConfig on every client creation.

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/service/ClusterProxyService.java (hunk 3 above)
Comment (sodre90) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/service/ClusterProxyService.java]: I asked CP guys and it is OK to call readConfig on every client creation

File: cluster-proxy/src/main/java/com/sequenceiq/cloudbreak/clusterproxy/ClusterProxyRegistrationClient.java [hunk 14]
```diff
@@ -30,6 +32,7 @@
         this.restTemplate = restTemplate;
     }
 
+    @Retryable(backoff = @Backoff(delay = 1000, multiplier = 2, maxDelay = 10000), maxAttempts = 5)
```
Comment (lacikaaa) [inline cluster-proxy/src/main/java/com/sequenceiq/cloudbreak/clusterproxy/ClusterProxyRegistrationClient.java:35]: nit: could you use constants for the values? it would be easier to adjust later

File: core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ClusterRepairService.java [hunk 15]
```diff
@@ -193,6 +195,21 @@ public FlowIdentifier repairNodes(Long stackId, Set<String> nodeIds, boolean del
         return repairStartResult;
     }
 
+    private boolean isPrimaryGWUnhealthyAndItIsNotSelected(ManualClusterRepairMode repairMode, Set<String> selectedParts, Stack stack) {
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ClusterRepairService.java:198]: could you add a log line to every branch so it would be easy to tell from the logs which returned and what

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java [hunk 16]
```diff
@@ -184,12 +164,114 @@ public int saveInstanceMetaData(Stack stack, Iterable<CloudVmMetaDataStatus> clo
...
         }
     }
 
+    private boolean setupInstanceMetaDataType(boolean primaryIgSelected, Optional<InstanceMetaData> terminatedPrimaryGwWhichShouldBeRestored,
+            InstanceMetaData instanceMetaDataEntry, InstanceGroup ig) {
+        if (instanceMetaDataEntry.getInstanceMetadataType() == null) {
+            if (ig != null) {
+                LOGGER.info("Instance group type: {}", ig.getInstanceGroupType());
+                if (InstanceGroupType.GATEWAY.equals(ig.getInstanceGroupType())) {
+                    primaryIgSelected = setupGatewayInstance(primaryIgSelected, terminatedPrimaryGwWhichShouldBeRestored, instanceMetaDataEntry);
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java:181]: I don't think it's a good idea to change the method parameter value here. either introduce a new variable and assign the initial value to the parameter or reorganize the code a bit and use returns at multiple lines

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java [hunk 17]
```diff
@@ -184,12 +164,114 @@ public int saveInstanceMetaData(Stack stack, Iterable<CloudVmMetaDataStatus> clo
...
+        }
+    }
+
+    private boolean setupGatewayInstance(boolean primaryIgSelected, Optional<InstanceMetaData> terminatedPrimaryGwWhichShouldBeRestored,
+            InstanceMetaData instanceMetaDataEntry) {
+        LOGGER.info("Instance is a gateway instance: {}", instanceMetaDataEntry.getInstanceId());
+        if (terminatedPrimaryGwWhichShouldBeRestored.isPresent()) {
+            boolean primaryGWRestored = restorePrimaryGWIfFQDNMatch(terminatedPrimaryGwWhichShouldBeRestored.get(), instanceMetaDataEntry);
+            if (primaryGWRestored) {
+                primaryIgSelected = true;
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java:220]: same here, please don't modify the parameter value, just insert return here
please add a log line

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java [hunk 18]
```diff
@@ -184,12 +164,114 @@ public int saveInstanceMetaData(Stack stack, Iterable<CloudVmMetaDataStatus> clo
...
+            }
+        } else {
+            primaryIgSelected = selectPrimaryGWIfNotSelected(primaryIgSelected, instanceMetaDataEntry);
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java:223]: also you could use return here

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java [hunk 19]
```diff
@@ -184,12 +164,114 @@ public int saveInstanceMetaData(Stack stack, Iterable<CloudVmMetaDataStatus> clo
...
+            }
+        } else {
+            LOGGER.error("There is no gateway group!");
+            throw new CloudbreakServiceException("There is no gateway group!");
+        }
+    }
+
+    private boolean selectPrimaryGWIfNotSelected(boolean primaryIgSelected, InstanceMetaData instanceMetaDataEntry) {
+        if (!primaryIgSelected) {
+            primaryIgSelected = true;
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java:252]: similar as before, just return the end of the if and else branch with the value

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java [hunk 20]
```diff
@@ -184,12 +164,114 @@ public int saveInstanceMetaData(Stack stack, Iterable<CloudVmMetaDataStatus> clo
...
+    private void selectPrimaryGWFromGatewayInstances(Collection<InstanceGroup> instanceGroups) {
+        Optional<InstanceGroup> gwInstanceGroup = instanceGroups.stream()
+                .filter(instanceGroup -> InstanceGroupType.GATEWAY.equals(instanceGroup.getInstanceGroupType()))
+                .findFirst();
+
+        if (gwInstanceGroup.isPresent()) {
+            List<InstanceMetaData> gwInstances = instanceMetaDataService.findAllByInstanceGroupAndInstanceStatus(gwInstanceGroup.get(), InstanceStatus.CREATED);
+            Optional<InstanceMetaData> gwInstance = gwInstances.stream().findFirst();
+            if (gwInstance.isPresent()) {
+                LOGGER.info("We were able to select primary GW from GW instances");
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java:237]: could you also log which was chosen?

**Review by lacikaaa — CHANGES_REQUESTED** [review]

**Review by lacikaaa — APPROVED** [review]
